platform/utm_mgrs_utils.cpp

```diff
@@ -0,0 +1,567 @@
+{
+  return to_string(point.zone_number) + string(1, point.zone_letter) + " " +\
+         to_string(int(round(point.easting))) + " " + \
+         to_string(int(round(point.northing)));
+}
+
+// Convert UTM point parameters to MGRS parameters. Additional 2 char code is deducted.
+// Easting and northing parameters are reduced to 5 digits.
+string utm_to_mgrs_str(UTMPoint point, int precision)
+{
+  if (point.zone_letter == 'Z')
+    return "Latitude limit exceeded";
+  else
+  {
+    string eastingStr = strings::to_string_width(point.easting, precision+1);
+    string northingStr = strings::to_string_width(point.northing, precision+1);
+
+    if (northingStr.size() > 6)
+      northingStr = northingStr.substr(northingStr.size() - 6);
+
+    return strings::to_string_width((long) point.zone_number, 2) + point.zone_letter + " " + \
+           get_100k_id(point.easting, point.northing, point.zone_number) + " " + \
+           eastingStr.substr(1, precision) + " " + \
+           northingStr.substr(1, precision);
+  }
+}
+
+// Build 2 chars string with 100k square ID
+string get_100k_id(double easting, double northing, int zone_number)
+{
+  int set = zone_to_100k(zone_number);
+  int setColumn = ((int) easting / 100000);
+  int setRow = ((int) northing / 100000) % 20;
+
+  int colOrigin = SET_ORIGIN_COLUMN_LETTERS[set - 1];
+  int rowOrigin = SET_ORIGIN_ROW_LETTERS[set - 1];
+
+  int colInt = colOrigin + setColumn - 1;
+  int rowInt = rowOrigin + setRow;
+  bool rollover = false;
+
+  if (colInt > 'Z')
+  {
+    colInt = colInt - 'Z' + 'A' - 1;
+    rollover = true;
+  }
+
+  if (colInt == 'I' || (colOrigin < 'I' && colInt > 'I') || ((colInt > 'I' || colOrigin < 'I') && rollover))
+    colInt++;
+  if (colInt == 'O' || (colOrigin < 'O' && colInt > 'O') || ((colInt > 'O' || colOrigin < 'O') && rollover))
+  {
+    colInt++;
+    if (colInt == 'I')
+      colInt++;
+  }
+
+  if (colInt > 'Z')
+    colInt = colInt - 'Z' + 'A' - 1;
+
+  if (rowInt > 'V')
+  {
+    rowInt = rowInt - 'V' + 'A' - 1;
+    rollover = true;
+  }
+  else
+    rollover = false;
+
+  if (rowInt == 'I' || (rowOrigin < 'I' && rowInt > 'I') || ((rowInt > 'I' || rowOrigin < 'I') && rollover)) {
+    rowInt++;
+  }
+
+  if (rowInt == 'O' || (rowOrigin < 'O' && rowInt > 'O') || ((rowInt > 'O' || rowOrigin < 'O') && rollover))
+  {
+    rowInt++;
+    if (rowInt == 'I')
+      rowInt++;
+  }
+
+  if (rowInt > 'V')
+    rowInt = rowInt - 'V' + 'A' - 1;
+
+  string twoLetter = {char(colInt), char(rowInt)};
+
+  return twoLetter;
+}
+
+int zone_to_100k(int i) {
+  int set = i % NUM_100K_SETS;
+  if (set == 0)
+    set = NUM_100K_SETS;
+  return set;
+}
+
+// Convert UTM parameters to lat,lon for WSG 84 ellipsoid.
+// If UTM parameters are valid lat and lon references are used to output calculated coordinates.
+// Otherwise function returns 'false'.
+bool UTMtoLatLon(double easting, double northing, int zone_number, char zone_letter, double &lat, double &lon)
```

Comment:
```suggestion
bool UTMtoLatLon(double easting, double northing, int zone_number, char zone_letter, double & lat, double & lon)
```

Can std::optional<ms::LatLon> be used for return type?

Reply:
Fixed formatting